File: src/main/java/com/jacuum/algo/impl/RandomAlgo.java
package com.jacuum.algo.impl;

import com.jacuum.algo.*;
import java.util.*;

/**
 * Picks uniformly at random from all passable directions (neighbors without walls).
 * When surrounded, picks from all four directions (robot stays in place).
 */
@RobotAlgorithm("Random")
public final class RandomAlgo implements RobotAlgo {
    private final Random rng;

    public RandomAlgo(final Random rng) {
        this.rng = rng;
    }

    @Override
    public Direction next(final Tile tile) {
        final List<Direction> passable = new ArrayList<>(4);
        for (final Direction d : Direction.values())
            if (!tile.hasWall(d)) passable.add(d);
        if (passable.isEmpty())
            return Direction.values()[rng.nextInt(Direction.values().length)];
        return passable.get(rng.nextInt(passable.size()));
    }
}

File: src/main/kotlin/com/jacuum/algo/impl/RandomAlgo.kt
package com.jacuum.algo.impl

import com.jacuum.algo.Direction
import com.jacuum.algo.RobotAlgo
import com.jacuum.algo.RobotAlgorithm
import com.jacuum.algo.Tile
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.annotation.Scope
import java.util.Random

@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@RobotAlgorithm("Random")
class RandomAlgo : RobotAlgo {
    private val rng = Random()
    override fun next(tile: Tile): Direction {
        val passable = Direction.entries.filter { !tile.hasWall(it) }
        val choices = passable.ifEmpty { Direction.entries }
        return choices[rng.nextInt(choices.size)]
    }
}
